File: frontend/src/utils/resilience.ts
import { defaultLogger } from './logger';
import { AppError, ErrorType, ErrorSeverity, errorHandler } from './errorHandler';

/**
 * 熔断器状态枚举
 */
export enum CircuitState {
  CLOSED = 'CLOSED',    // 正常状态，允许请求通过
  OPEN = 'OPEN',        // 熔断状态，拒绝所有请求
  HALF_OPEN = 'HALF_OPEN'  // 半开状态，允许部分请求通过以测试服务是否恢复
}

/**
 * 熔断器配置接口
 */
export interface CircuitBreakerConfig {
  failureThreshold: number;     // 触发熔断的失败次数阈值
  recoveryTimeout: number;      // 熔断恢复超时时间（毫秒）
  halfOpenMaxCalls: number;     // 半开状态下允许的最大请求数
}

/**
 * 重试配置接口
 */
export interface RetryConfig {
  maxRetries: number;           // 最大重试次数
  initialBackoff: number;       // 初始退避时间（毫秒）
  maxBackoff: number;           // 最大退避时间（毫秒）
  backoffMultiplier: number;    // 退避时间乘数
  jitter: boolean;              // 是否添加随机抖动
}

/**
 * 熔断器类
 * 用于防止对失败服务的持续请求，避免级联故障
 */
export class CircuitBreaker {
  private name: string;
  private config: CircuitBreakerConfig;
  private state: CircuitState = CircuitState.CLOSED;
  private failureCount: number = 0;
  private lastFailureTime: number = 0;
  private halfOpenCalls: number = 0;

  constructor(name: string, config?: Partial<CircuitBreakerConfig>) {
    this.name = name;
    this.config = {
      failureThreshold: config?.failureThreshold || 5,
      recoveryTimeout: config?.recoveryTimeout || 30000,
      halfOpenMaxCalls: config?.halfOpenMaxCalls || 3
    };
  }

  /**
   * 检查是否允许请求通过
   */
  public allowRequest(): boolean {
    const currentTime = Date.now();

    if (this.state === CircuitState.OPEN) {
      // 检查是否达到恢复超时时间
      if (currentTime - this.lastFailureTime >= this.config.recoveryTimeout) {
        defaultLogger.info(`熔断器 ${this.name} 进入半开状态`);
        this.state = CircuitState.HALF_OPEN;
        this.halfOpenCalls = 0;
        return true;
      }
      return false;
    }

    if (this.state === CircuitState.HALF_OPEN) {
      // 在半开状态下限制请求数量
      if (this.halfOpenCalls < this.config.halfOpenMaxCalls) {
        this.halfOpenCalls += 1;
        return true;
      }
      return false;
    }

    // 闭合状态，允许所有请求
    return true;
  }

  /**
   * 记录成功请求
   */
  public recordSuccess(): void {
    if (this.state === CircuitState.HALF_OPEN) {
      defaultLogger.info(`熔断器 ${this.name} 恢复正常，关闭熔断`);
      this.state = CircuitState.CLOSED;
      this.failureCount = 0;
      this.halfOpenCalls = 0;
    }
  }

  /**
   * 记录失败请求
   */
  public recordFailure(): void {
    const currentTime = Date.now();
    this.lastFailureTime = currentTime;

    if (this.state === CircuitState.HALF_OPEN) {
      defaultLogger.warn(`熔断器 ${this.name} 在半开状态下检测到失败，重新打开熔断`);
      this.state = CircuitState.OPEN;
      return;
    }

    this.failureCount += 1;

    if (this.state === CircuitState.CLOSED && this.failureCount >= this.config.failureThreshold) {
      defaultLogger.warn(`熔断器 ${this.name} 失败次数达到阈值 ${this.config.failureThreshold}，打开熔断`);
      this.state = CircuitState.OPEN;
    }
  }

  /**
   * 获取当前熔断器状态
   */
  public getState(): CircuitState {
    return this.state;
  }

  /**
   * 重置熔断器状态
   */
  public reset(): void {
    this.state = CircuitState.CLOSED;
    this.failureCount = 0;
    this.lastFailureTime = 0;
    this.halfOpenCalls = 0;
  }
}

/**
 * 熔断器注册表
 * 用于管理和复用熔断器实例
 */
class CircuitBreakerRegistry {
  private static instance: CircuitBreakerRegistry;
  private breakers: Map<string, CircuitBreaker> = new Map();

  private constructor() {}

  public static getInstance(): CircuitBreakerRegistry {
    if (!CircuitBreakerRegistry.instance) {
      CircuitBreakerRegistry.instance = new CircuitBreakerRegistry();
    }
    return CircuitBreakerRegistry.instance;
  }

  /**
   * 获取或创建熔断器
   */
  public getBreaker(name: string, config?: Partial<CircuitBreakerConfig>): CircuitBreaker {
    if (!this.breakers.has(name)) {
      this.breakers.set(name, new CircuitBreaker(name, config));
    }
    return this.breakers.get(name)!;
  }

  /**
   * 重置所有熔断器
   */
  public resetAll(): void {
    this.breakers.forEach(breaker => breaker.reset());
  }
}

// 导出熔断器注册表单例
export const circuitBreakerRegistry = CircuitBreakerRegistry.getInstance();

/**
 * 重试工具类
 * 提供自动重试功能，支持指数退避策略
 */
export class RetryUtil {
  /**
   * 使用重试策略执行异步函数
   * @param fn 要执行的异步函数
   * @param config 重试配置
   * @param breakerName 熔断器名称（可选）
   * @returns 函数执行结果的Promise
   */
  public static async withRetry<T>(
    fn: () => Promise<T>,
    config?: Partial<RetryConfig>,
    breakerName?: string
  ): Promise<T> {
    const retryConfig: RetryConfig = {
      maxRetries: config?.maxRetries || 3,
      initialBackoff: config?.initialBackoff || 1000,
      maxBackoff: config?.maxBackoff || 30000,
      backoffMultiplier: config?.backoffMultiplier || 2,
      jitter: config?.jitter !== undefined ? config.jitter : true
    };

    let breaker: CircuitBreaker | undefined;
    if (breakerName) {
      breaker = circuitBreakerRegistry.getBreaker(breakerName);
    }

    let retryCount = 0;
    let lastError: any;

    while (retryCount <= retryConfig.maxRetries) {
      try {
        // 检查熔断器状态
        if (breaker && !breaker.allowRequest()) {
          const error = new AppError(
            `服务 ${breakerName} 暂时不可用`,
            ErrorType.NETWORK,
            ErrorSeverity.ERROR,
            { breakerName },
            '服务暂时不可用，请稍后重试'
          );
          errorHandler.handleError(error);
          throw error;
        }

        // 执行函数
        const result = await fn();

        // 记录成功
        if (breaker) {
          breaker.recordSuccess();
        }

        return result;
      } catch (error) {
        lastError = error;
        retryCount += 1;

        // 记录失败
        if (breaker) {
          breaker.recordFailure();
        }

        // 达到最大重试次数，抛出异常
        if (retryCount > retryConfig.maxRetries) {
          defaultLogger.error(`重试次数达到上限 ${retryConfig.maxRetries}，放弃重试`, {
            function: fn.name || 'anonymous',
            error,
            retryCount
          });
          throw error;
        }

        // 计算退避时间
        let backoff = Math.min(
          retryConfig.initialBackoff * Math.pow(retryConfig.backoffMultiplier, retryCount - 1),
          retryConfig.maxBackoff
        );

        // 添加抖动以避免惊群效应
        if (retryConfig.jitter) {
          backoff = backoff * (0.5 + Math.random());
        }

        defaultLogger.warn(`操作失败，将在 ${(backoff / 1000).toFixed(2)} 秒后重试 (${retryCount}/${retryConfig.maxRetries})`, {
          function: fn.name || 'anonymous',
          error,
          retryCount,
          backoff
        });

        // 等待退避时间
        await new Promise(resolve => setTimeout(resolve, backoff));
      }
    }

    // 不应该到达这里，但为了类型安全添加
    throw lastError;
  }

  /**
   * 创建带有重试功能的API调用包装器
   * @param apiClient API客户端对象
   * @param config 重试配置
   * @returns 包装后的API客户端
   */
  public static createRetryableApiClient<T extends object>(
    apiClient: T,
    config?: Partial<RetryConfig>
  ): T {
    const retryableClient = {} as T;

    // 遍历原始客户端的所有方法
    for (const key of Object.getOwnPropertyNames(Object.getPrototypeOf(apiClient))) {
      const prop = (apiClient as any)[key];

      // 只包装函数
      if (typeof prop === 'function' && key !== 'constructor') {
        (retryableClient as any)[key] = async (...args: any[]) => {
          return RetryUtil.withRetry(
            () => prop.apply(apiClient, args),
            config,
            `api.${key}`
          );
        };
      } else {
        (retryableClient as any)[key] = prop;
      }
    }

    return retryableClient;
  }
}

/**
 * 超时控制工具类
 */
export class TimeoutUtil {
  /**
   * 使用超时控制执行异步函数
   * @param fn 要执行的异步函数
   * @param timeoutMs 超时时间（毫秒）
   * @returns 函数执行结果的Promise
   */
  public static async withTimeout<T>(fn: () => Promise<T>, timeoutMs: number): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      // 创建超时计时器
      const timeoutId = setTimeout(() => {
        const error = new AppError(
          `操作超时 (${timeoutMs}毫秒)`,
          ErrorType.NETWORK,
          ErrorSeverity.ERROR,
          { timeoutMs },
          '操作超时，请稍后重试'
        );
        reject(error);
      }, timeoutMs);

      // 执行函数
      fn()
        .then(result => {
          clearTimeout(timeoutId);
          resolve(result);
        })
        .catch(error => {
          clearTimeout(timeoutId);
          reject(error);
        });
    });
  }
}

/**
 * 限流工具类
 * 用于控制并发请求数量
 */
export class RateLimiter {
  private maxConcurrent: number;
  private currentConcurrent: number = 0;
  private queue: Array<() => void> = [];

  constructor(maxConcurrent: number = 5) {
    this.maxConcurrent = maxConcurrent;
  }

  /**
   * 使用限流控制执行异步函数
   * @param fn 要执行的异步函数
   * @returns 函数执行结果的Promise
   */
  public async execute<T>(fn: () => Promise<T>): Promise<T> {
    // 等待获取执行权
    await this.acquireSlot();

    try {
      // 执行函数
      return await fn();
    } finally {
      // 释放执行权
      this.releaseSlot();
    }
  }

  /**
   * 获取执行槽位
   */
  private async acquireSlot(): Promise<void> {
    if (this.currentConcurrent < this.maxConcurrent) {
      this.currentConcurrent += 1;
      return;
    }

    // 如果已达到最大并发数，则等待
    return new Promise<void>(resolve => {
      this.queue.push(resolve);
    });
  }

  /**
   * 释放执行槽位
   */
  private releaseSlot(): void {
    if (this.queue.length > 0) {
      // 如果队列中有等待的任务，则唤醒一个
      const next = this.queue.shift();
      if (next) {
        next();
      }
    } else {
      // 否则减少当前并发数
      this.currentConcurrent -= 1;
    }
  }
}

File: backend/app/core/resilience.py
import asyncio
import functools
import time
from enum import Enum
from typing import Any, Callable, Dict, List, Optional, TypeVar, Union, cast
from app.core.logger import logger

# 定义类型变量
T = TypeVar('T')
F = TypeVar('F', bound=Callable[..., Any])

class CircuitState(str, Enum):
    """熔断器状态枚举"""
    CLOSED = "CLOSED"  # 正常状态，允许请求通过
    OPEN = "OPEN"      # 熔断状态，拒绝所有请求
    HALF_OPEN = "HALF_OPEN"  # 半开状态，允许部分请求通过以测试服务是否恢复


class CircuitBreaker:
    """
    熔断器实现
    当失败率达到阈值时，熔断器打开，阻止请求通过，防止级联失败
    """
    
    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        recovery_timeout: float = 30.0,
        half_open_max_calls: int = 3
    ):
        self.name = name
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.half_open_max_calls = half_open_max_calls
        
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.last_failure_time = 0.0
        self.half_open_calls = 0
    
    def allow_request(self) -> bool:
        """检查是否允许请求通过"""
        current_time = time.time()
        
        if self.state == CircuitState.OPEN:
            # 检查是否达到恢复超时时间
            if current_time - self.last_failure_time >= self.recovery_timeout:
                logger.info(f"熔断器 {self.name} 进入半开状态")
                self.state = CircuitState.HALF_OPEN
                self.half_open_calls = 0
                return True
            return False
        
        elif self.state == CircuitState.HALF_OPEN:
            # 在半开状态下限制请求数量
            if self.half_open_calls < self.half_open_max_calls:
                self.half_open_calls += 1
                return True
            return False
        
        # 闭合状态，允许所有请求
        return True
    
    def record_success(self) -> None:
        """记录成功请求"""
        if self.state == CircuitState.HALF_OPEN:
            logger.info(f"熔断器 {self.name} 恢复正常，关闭熔断")
            self.state = CircuitState.CLOSED
            self.failure_count = 0
            self.half_open_calls = 0
    
    def record_failure(self) -> None:
        """记录失败请求"""
        current_time = time.time()
        self.last_failure_time = current_time
        
        if self.state == CircuitState.HALF_OPEN:
            logger.warning(f"熔断器 {self.name} 在半开状态下检测到失败，重新打开熔断")
            self.state = CircuitState.OPEN
            return
        
        self.failure_count += 1
        
        if self.state == CircuitState.CLOSED and self.failure_count >= self.failure_threshold:
            logger.warning(f"熔断器 {self.name} 失败次数达到阈值 {self.failure_threshold}，打开熔断")
            self.state = CircuitState.OPEN


class RetryConfig:
    """重试配置"""
    
    def __init__(
        self,
        max_retries: int = 3,
        initial_backoff: float = 1.0,
        max_backoff: float = 30.0,
        backoff_multiplier: float = 2.0,
        jitter: bool = True
    ):
        self.max_retries = max_retries
        self.initial_backoff = initial_backoff
        self.max_backoff = max_backoff
        self.backoff_multiplier = backoff_multiplier
        self.jitter = jitter


class ResilienceDecorators:
    """弹性装饰器集合，提供重试和熔断功能"""
    
    _circuit_breakers: Dict[str, CircuitBreaker] = {}
    
    @classmethod
    def get_circuit_breaker(cls, name: str) -> CircuitBreaker:
        """获取或创建熔断器实例"""
        if name not in cls._circuit_breakers:
            cls._circuit_breakers[name] = CircuitBreaker(name)
        return cls._circuit_breakers[name]
    
    @classmethod
    def retry(
        cls,
        retryable_exceptions: List[type] = None,
        retry_config: RetryConfig = None,
        circuit_breaker_name: Optional[str] = None
    ):
        """
        重试装饰器
        在遇到指定异常时自动重试函数调用
        """
        retryable_exceptions = retryable_exceptions or [Exception]
        retry_config = retry_config or RetryConfig()
        
        def decorator(func: F) -> F:
            @functools.wraps(func)
            async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
                circuit_breaker = None
                if circuit_breaker_name:
                    circuit_breaker = cls.get_circuit_breaker(circuit_breaker_name)
                
                retry_count = 0
                last_exception = None
                
                while retry_count <= retry_config.max_retries:
                    try:
                        # 检查熔断器状态
                        if circuit_breaker and not circuit_breaker.allow_request():
                            logger.warning(f"熔断器 {circuit_breaker_name} 已打开，拒绝请求")
                            raise Exception(f"服务 {circuit_breaker_name} 暂时不可用，请稍后重试")
                        
                        # 执行函数
                        result = await func(*args, **kwargs)
                        
                        # 记录成功
                        if circuit_breaker:
                            circuit_breaker.record_success()
                        
                        return result
                    
                    except tuple(retryable_exceptions) as e:
                        last_exception = e
                        retry_count += 1
                        
                        # 记录失败
                        if circuit_breaker:
                            circuit_breaker.record_failure()
                        
                        # 达到最大重试次数，抛出异常
                        if retry_count > retry_config.max_retries:
                            logger.error(f"重试次数达到上限 {retry_config.max_retries}，放弃重试", extra={
                                "function": func.__name__,
                                "exception": str(e),
                                "retry_count": retry_count
                            })
                            raise
                        
                        # 计算退避时间
                        backoff = min(
                            retry_config.initial_backoff * (retry_config.backoff_multiplier ** (retry_count - 1)),
                            retry_config.max_backoff
                        )
                        
                        # 添加抖动以避免惊群效应
                        if retry_config.jitter:
                            import random
                            backoff = backoff * (0.5 + random.random())
                        
                        logger.warning(f"操作失败，将在 {backoff:.2f} 秒后重试 ({retry_count}/{retry_config.max_retries})", extra={
                            "function": func.__name__,
                            "exception": str(e),
                            "retry_count": retry_count,
                            "backoff": backoff
                        })
                        
                        # 等待退避时间
                        await asyncio.sleep(backoff)
                
                # 不应该到达这里，但为了类型检查添加
                assert last_exception is not None
                raise last_exception
            
            @functools.wraps(func)
            def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
                circuit_breaker = None
                if circuit_breaker_name:
                    circuit_breaker = cls.get_circuit_breaker(circuit_breaker_name)
                
                retry_count = 0
                last_exception = None
                
                while retry_count <= retry_config.max_retries:
                    try:
                        # 检查熔断器状态
                        if circuit_breaker and not circuit_breaker.allow_request():
                            logger.warning(f"熔断器 {circuit_breaker_name} 已打开，拒绝请求")
                            raise Exception(f"服务 {circuit_breaker_name} 暂时不可用，请稍后重试")
                        
                        # 执行函数
                        result = func(*args, **kwargs)
                        
                        # 记录成功
                        if circuit_breaker:
                            circuit_breaker.record_success()
                        
                        return result
                    
                    except tuple(retryable_exceptions) as e:
                        last_exception = e
                        retry_count += 1
                        
                        # 记录失败
                        if circuit_breaker:
                            circuit_breaker.record_failure()
                        
                        # 达到最大重试次数，抛出异常
                        if retry_count > retry_config.max_retries:
                            logger.error(f"重试次数达到上限 {retry_config.max_retries}，放弃重试", extra={
                                "function": func.__name__,
                                "exception": str(e),
                                "retry_count": retry_count
                            })
                            raise
                        
                        # 计算退避时间
                        backoff = min(
                            retry_config.initial_backoff * (retry_config.backoff_multiplier ** (retry_count - 1)),
                            retry_config.max_backoff
                        )
                        
                        # 添加抖动以避免惊群效应
                        if retry_config.jitter:
                            import random
                            backoff = backoff * (0.5 + random.random())
                        
                        logger.warning(f"操作失败，将在 {backoff:.2f} 秒后重试 ({retry_count}/{retry_config.max_retries})", extra={
                            "function": func.__name__,
                            "exception": str(e),
                            "retry_count": retry_count,
                            "backoff": backoff
                        })
                        
                        # 等待退避时间
                        time.sleep(backoff)
                
                # 不应该到达这里，但为了类型检查添加
                assert last_exception is not None
                raise last_exception
            
            # 根据函数是否为协程函数选择包装器
            if asyncio.iscoroutinefunction(func):
                return cast(F, async_wrapper)
            else:
                return cast(F, sync_wrapper)
        
        return decorator
    
    @classmethod
    def circuit_breaker(cls, name: str, failure_threshold: int = 5, recovery_timeout: float = 30.0):
        """
        熔断器装饰器
        当失败次数达到阈值时，阻止请求通过一段时间
        """
        def decorator(func: F) -> F:
            circuit_breaker = cls.get_circuit_breaker(name)
            circuit_breaker.failure_threshold = failure_threshold
            circuit_breaker.recovery_timeout = recovery_timeout
            
            @functools.wraps(func)
            async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
                if not circuit_breaker.allow_request():
                    logger.warning(f"熔断器 {name} 已打开，拒绝请求")
                    raise Exception(f"服务 {name} 暂时不可用，请稍后重试")
                
                try:
                    result = await func(*args, **kwargs)
                    circuit_breaker.record_success()
                    return result
                except Exception as e:
                    circuit_breaker.record_failure()
                    raise
            
            @functools.wraps(func)
            def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
                if not circuit_breaker.allow_request():
                    logger.warning(f"熔断器 {name} 已打开，拒绝请求")
                    raise Exception(f"服务 {name} 暂时不可用，请稍后重试")
                
                try:
                    result = func(*args, **kwargs)
                    circuit_breaker.record_success()
                    return result
                except Exception as e:
                    circuit_breaker.record_failure()
                    raise
            
            # 根据函数是否为协程函数选择包装器
            if asyncio.iscoroutinefunction(func):
                return cast(F, async_wrapper)
            else:
                return cast(F, sync_wrapper)
        
        return decorator
    
    @classmethod
    def timeout(cls, seconds: float):
        """
        超时装饰器
        在指定时间后取消操作
        """
        def decorator(func: F) -> F:
            @functools.wraps(func)
            async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
                try:
                    return await asyncio.wait_for(func(*args, **kwargs), timeout=seconds)
                except asyncio.TimeoutError:
                    logger.error(f"操作超时 ({seconds}秒)", extra={
                        "function": func.__name__,
                        "timeout": seconds
                    })
                    raise Exception(f"操作超时，请稍后重试")
            
            @functools.wraps(func)
            def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
                import signal
                
                def timeout_handler(signum, frame):
                    raise Exception(f"操作超时 ({seconds}秒)")
                
                # 设置信号处理器
                original_handler = signal.signal(signal.SIGALRM, timeout_handler)
                signal.alarm(int(seconds))
                
                try:
                    return func(*args, **kwargs)
                finally:
                    # 恢复原始信号处理器
                    signal.alarm(0)
                    signal.signal(signal.SIGALRM, original_handler)
            
            # 根据函数是否为协程函数选择包装器
            if asyncio.iscoroutinefunction(func):
                return cast(F, async_wrapper)
            else:
                return cast(F, sync_wrapper)
        
        return decorator


# 导出便捷函数
retry = ResilienceDecorators.retry
circuit_breaker = ResilienceDecorators.circuit_breaker
timeout = ResilienceDecorators.timeout